packages/flutter/lib/src/cupertino/nav_bar.dart

```diff
@@ -150,18 +313,67 @@ class CupertinoNavigationBar extends StatelessWidget implements ObstructingPrefe
     return const Size.fromHeight(_kNavBarPersistentHeight);
   }
 
+  @override
+  _CupertinoNavigationBarState createState() {
+    return new _CupertinoNavigationBarState();
+  }
+}
+
+// A state class exists for the nav bar so that the keys of its sub-components
+// don't change when rebuilding the nav bar, causing the sub-components to
+// lose their own states.
+class _CupertinoNavigationBarState extends State<CupertinoNavigationBar> {
+  _NavigationBarStaticComponentsKeys keys;
+
+  @override
+  void initState() {
+    super.initState();
+    keys = new _NavigationBarStaticComponentsKeys();
+  }
+
   @override
   Widget build(BuildContext context) {
-    return _wrapWithBackground(
-      border: border,
-      backgroundColor: backgroundColor,
-      child: new _CupertinoPersistentNavigationBar(
-        leading: leading,
-        automaticallyImplyLeading: automaticallyImplyLeading,
-        middle: new Semantics(child: middle, header: true),
-        trailing: trailing,
-        padding: padding,
-        actionsForegroundColor: actionsForegroundColor,
+    final _NavigationBarStaticComponents components = new _NavigationBarStaticComponents(
+      keys: keys,
+      route: ModalRoute.of(context),
+      userLeading: widget.leading,
+      automaticallyImplyLeading: widget.automaticallyImplyLeading,
+      automaticallyImplyTitle: widget.automaticallyImplyMiddle,
+      previousPageTitle: widget.previousPageTitle,
+      userMiddle: widget.middle,
+      userTrailing: widget.trailing,
+      padding: widget.padding,
+      actionsForegroundColor: widget.actionsForegroundColor,
+      userLargeTitle: null,
+      large: false,
+    );
+
+    final Widget navBar = _wrapWithBackground(
+      border: widget.border,
+      backgroundColor: widget.backgroundColor,
+      child: new _PersistentNavigationBar(
+        components: components,
+        padding: widget.padding,
+      ),
+    );
+
+    if (!widget.transitionBetweenRoutes || !_isTransitionable(context)) {
+      return navBar;
+    }
+
+    return new Hero(
+      tag: _heroTag,
```

Comment:
We should follow the pattern of `FloatingActionButton.heroTag` to allow people to have multiple nav bars at once.

Reply:
Umm, interesting. Sure